In [1]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
accidents = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
accidents.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [6]:
accidents.isnull()

Index  Accident_Severity  Accident Date  Latitude  Light_Conditions  \
0       False              False          False     False             False   
1       False              False          False     False             False   
2       False              False          False     False             False   
3       False              False          False     False             False   
4       False              False          False     False             False   
...       ...                ...            ...       ...               ...   
660674  False              False          False     False             False   
660675  False              False          False     False             False   
660676  False              False          False     False             False   
660677  False              False          False     False             False   
660678  False              False          False     False             False   

        District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0               False      False                 False               False   
1               False      False                 False               False   
2               False      False                 False               False   
3               False      False                 False               False   
4               False      False                 False               False   
...               ...        ...                   ...                 ...   
660674          False      False                 False               False   
660675          False      False                 False               False   
660676          False      False                 False               False   
660677          False      False                 False               False   
660678          False      False                 False               False   

        Road_Surface_Conditions  Road_Type  Urban_or_Rural_Area  \
0                         False      False                False   
1                         False      False                False   
2                         False       True                False   
3                         False      False                False   
4                         False       True                False   
...                         ...        ...                  ...   
660674                    False      False                False   
660675                    False      False                False   
660676                    False      False                False   
660677                    False      False                False   
660678                    False      False                False   

        Weather_Conditions  Vehicle_Type  
0                    False         False  
1                    False         False  
2                     True         False  
3                    False         False  
4                     True         False  
...                    ...           ...  
660674               False         False  
660675               False         False  
660676               False         False  
660677               False         False  
660678               False         False  

[660679 rows x 14 columns]

In [7]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [8]:
accidents['Latitude']=accidents['Latitude'].fillna(accidents['Latitude'].mode()[0])

In [9]:
accidents['Longitude']=accidents['Longitude'].fillna(accidents['Longitude'].mode()[0])

In [10]:
accidents['Road_Surface_Conditions']=accidents['Road_Surface_Conditions'].fillna(accidents['Road_Surface_Conditions'].mode()[0])

In [11]:
accidents['Road_Type']=accidents['Road_Type'].fillna(accidents['Road_Type'].mode()[0])

In [12]:
accidents['Urban_or_Rural_Area']=accidents['Urban_or_Rural_Area'].fillna(accidents['Longitude'].mode()[0])

In [13]:
accidents.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [14]:
accidents.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [15]:
accidents['Accident Date'] = accidents['Accident Date'].astype('str')

In [16]:
accidents['Accident Date'] = accidents['Accident Date'].str.strip()
accidents['Accident Date'] = accidents['Accident Date'].str.replace('/','-')

In [17]:
accidents['Accident Date'] = pd.to_datetime(accidents['Accident Date'], dayfirst = True, errors = 'coerce')

In [18]:
accidents.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

In [19]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                      0
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [20]:
accidents['year']=accidents['Accident Date'].dt.year

In [21]:
accidents['month']=accidents['Accident Date'].dt.month

In [22]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                      0
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
year                           0
month                          0
dtype: int64

<h1>QUESTION 1</h1>
<H3>Accident severity based on light conditions</H3>

In [23]:
severity1 = accidents.groupby(['Accident_Severity','Light_Conditions']).size()

In [24]:
severity1.unstack()

Light_Conditions   Darkness - lighting unknown  Darkness - lights lit  \
Accident_Severity                                                       
Fatal                                       68                   1860   
Serious                                    794                  19130   
Slight                                    5622                 108345   

Light_Conditions   Darkness - lights unlit  Darkness - no lighting  Daylight  
Accident_Severity                                                             
Fatal                                   45                    1612      5076  
Serious                                360                    7174     60759  
Slight                                2138                   28651    419045

<h1>Insight 1</h1>
<h3>There are more fatal accidents in daylight with 5,076, serious accidents are more occuring in the darkness where lights are lit, and slight damage occurs in daylight</h3>

<h1>QUESTION 2</h1>
<H3>Accident severity based on district area, what area has the most fatal accidents?</H3>

In [25]:
severity2 = accidents.groupby(['Accident_Severity','District Area']).size()

In [26]:
severity2.unstack()

District Area      Aberdeen City  Aberdeenshire  Adur  Allerdale  Alnwick  \
Accident_Severity                                                           
Fatal                         12             66     8         24        6   
Serious                      239            463   101        143       33   
Slight                      1072           1401   510        961      193   

District Area      Amber Valley  Angus  Argyll and Bute  Arun  Ashfield  ...  \
Accident_Severity                                                        ...   
Fatal                        19     32               27    15        11  ...   
Serious                     173    163              184   253       201  ...   
Slight                     1155    601              625  1108      1183  ...   

District Area      Wokingham  Wolverhampton  Worcester  Worthing  Wrexham  \
Accident_Severity                                                           
Fatal                     13             20          3         4       14   
Serious                  167            282         74       140      148   
Slight                  1130           2164        781       864     1140   

District Area      Wychavon  Wycombe  Wyre  Wyre Forest  York  
Accident_Severity                                              
Fatal                    30       20    15           22    21  
Serious                 193      216   186          132   255  
Slight                 1138     1493  1037          815  1621  

[3 rows x 422 columns]

<h1>Insight 2</h1>
<h3>There are more fatal and serious accidents in the district of Aberdeenshire</h3>

<h1>QUESTION 3</h1>
<H3>Accident severity based on road surface conditions</H3>

In [27]:
severity3 = accidents.groupby(['Accident_Severity','Road_Surface_Conditions']).size()

In [28]:
severity3.unstack()

Road_Surface_Conditions     Dry  Flood over 3cm. deep  Frost or ice  Snow  \
Accident_Severity                                                           
Fatal                      5790                    23           193    35   
Serious                   61708                   152          2007   565   
Slight                   381049                   842         16317  5290   

Road_Surface_Conditions  Wet or damp  
Accident_Severity                     
Fatal                           2620  
Serious                        23785  
Slight                        160303

<h1>Insight 3</h1>
<h3>There are more fatal and serious accidents in dry road surface conditions</h3>

<h1>QUESTION 4</h1>
<H3>Accident severity based on area of loacation(Rural or Urban)

In [29]:
severity4 = accidents.groupby(['Accident_Severity','Urban_or_Rural_Area']).size()

In [30]:
severity4.unstack()

Urban_or_Rural_Area  -0.977611     Rural   Unallocated     Urban
Accident_Severity                                               
Fatal                      NaN    5601.0           NaN    3060.0
Serious                    2.0   37312.0           1.0   50902.0
Slight                    13.0  196077.0          10.0  367701.0

<h1>Insight 4</h1>
<h3>There are more fatal accidents in rural areas, serious and slight accidents on Urban areas</h3>

<h1>QUESTION 5</h1>
<H3>Accident severity based on vehicle type, what type of vehicle is likely to have a fatal accident?</H3>

In [31]:
severity5 = accidents.groupby(['Accident_Severity','Vehicle_Type']).size()

In [32]:
severity5.unstack()

Vehicle_Type       Agricultural vehicle  Bus or coach (17 or more pass seats)  \
Accident_Severity                                                               
Fatal                              21.0                                 325.0   
Serious                           282.0                                3373.0   
Slight                           1644.0                               22180.0   

Vehicle_Type            Car  Data missing or out of range  \
Accident_Severity                                           
Fatal                6577.0                           NaN   
Serious             66461.0                           NaN   
Slight             424954.0                           6.0   

Vehicle_Type       Goods 7.5 tonnes mgw and over  \
Accident_Severity                                  
Fatal                                      216.0   
Serious                                   2321.0   
Slight                                   14770.0   

Vehicle_Type       Goods over 3.5t. and under 7.5t  \
Accident_Severity                                    
Fatal                                         67.0   
Serious                                      857.0   
Slight                                      5172.0   

Vehicle_Type       Minibus (8 - 16 passenger seats)  \
Accident_Severity                                     
Fatal                                          29.0   
Serious                                       276.0   
Slight                                       1671.0   

Vehicle_Type       Motorcycle 125cc and under  Motorcycle 50cc and under  \
Accident_Severity                                                          
Fatal                                   189.0                       95.0   
Serious                                2031.0                     1014.0   
Slight                                13049.0                     6494.0   

Vehicle_Type       Motorcycle over 125cc and up to 500cc  \
Accident_Severity                                          
Fatal                                              105.0   
Serious                                           1014.0   
Slight                                            6537.0   

Vehicle_Type       Motorcycle over 500cc  Other vehicle  Pedal cycle  \
Accident_Severity                                                      
Fatal                              339.0           70.0          6.0   
Serious                           3457.0          767.0         39.0   
Slight                           21861.0         4800.0        152.0   

Vehicle_Type       Ridden horse  Taxi/Private hire car  \
Accident_Severity                                        
Fatal                       NaN                  155.0   
Serious                     NaN                 1771.0   
Slight                      4.0                11368.0   

Vehicle_Type       Van / Goods 3.5 tonnes mgw or under  
Accident_Severity                                       
Fatal                                            467.0  
Serious                                         4554.0  
Slight                                         29139.0

<h1>Insight 5</h1><h3>There is a high chance that cars can have fatal accidents, followed by motorcyles and vans</h3>

<h1>QUESTION 6</h1>
<H3>Number of accident severity</H3>

In [33]:
accidents['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h1>Insight 6</h1><h3>Accident severity has the following values: Slight     563,801
Serious     88,217
Fatal        8,661</h3>

<h1>QUESTION 7</h1>
<H3>Most common roadtype accident?</H3>

In [34]:
accidents['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h1>Insight 7</h1><h3>There have been 75.17% of accidents recorded in Single carriageway type of road</h3>

<h1>QUESTION 8</h1>
<H3>Accident by weather condition?</H3>

In [35]:
accidents['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h1>Insight 8</h1><h3>There have been 80.56% of accidents recorded in Fine no high winds type of weather condition</h3>

<h1>QUESTION 9</h1>
<H3>Accident based on vehicle type? how severe is it?</H3>

In [36]:
accidents.groupby(['Accident_Severity','Vehicle_Type']).size().unstack()

Vehicle_Type       Agricultural vehicle  Bus or coach (17 or more pass seats)  \
Accident_Severity                                                               
Fatal                              21.0                                 325.0   
Serious                           282.0                                3373.0   
Slight                           1644.0                               22180.0   

Vehicle_Type            Car  Data missing or out of range  \
Accident_Severity                                           
Fatal                6577.0                           NaN   
Serious             66461.0                           NaN   
Slight             424954.0                           6.0   

Vehicle_Type       Goods 7.5 tonnes mgw and over  \
Accident_Severity                                  
Fatal                                      216.0   
Serious                                   2321.0   
Slight                                   14770.0   

Vehicle_Type       Goods over 3.5t. and under 7.5t  \
Accident_Severity                                    
Fatal                                         67.0   
Serious                                      857.0   
Slight                                      5172.0   

Vehicle_Type       Minibus (8 - 16 passenger seats)  \
Accident_Severity                                     
Fatal                                          29.0   
Serious                                       276.0   
Slight                                       1671.0   

Vehicle_Type       Motorcycle 125cc and under  Motorcycle 50cc and under  \
Accident_Severity                                                          
Fatal                                   189.0                       95.0   
Serious                                2031.0                     1014.0   
Slight                                13049.0                     6494.0   

Vehicle_Type       Motorcycle over 125cc and up to 500cc  \
Accident_Severity                                          
Fatal                                              105.0   
Serious                                           1014.0   
Slight                                            6537.0   

Vehicle_Type       Motorcycle over 500cc  Other vehicle  Pedal cycle  \
Accident_Severity                                                      
Fatal                              339.0           70.0          6.0   
Serious                           3457.0          767.0         39.0   
Slight                           21861.0         4800.0        152.0   

Vehicle_Type       Ridden horse  Taxi/Private hire car  \
Accident_Severity                                        
Fatal                       NaN                  155.0   
Serious                     NaN                 1771.0   
Slight                      4.0                11368.0   

Vehicle_Type       Van / Goods 3.5 tonnes mgw or under  
Accident_Severity                                       
Fatal                                            467.0  
Serious                                         4554.0  
Slight                                         29139.0

<h1>Insight 9</h1><h3>There have been 75.37% of accidents recorded in a Car</h3>

<h1>Insight 10</h1><h3>But based on recorded data in all types of severity of accident the leading vehicle is an Agricultural vehicle</h3>

<h1>QUESTION 10</h1>
<H3>Number of Casualties? How about per District?What District has the most casualty?least?</H3>

In [37]:
accidents['Number_of_Casualties'].sum()

np.int64(896568)

<h1>Insight 11</h1><h3>There have been a total of 896,568 casualties</h3>

In [38]:
accidents.groupby(['Number_of_Casualties','District Area']).size()

Number_of_Casualties  District Area          
1                     Aberdeen City              1181
                      Aberdeenshire              1531
                      Adur                        478
                      Allerdale                   805
                      Alnwick                     156
                                                 ... 
45                    Stratford-upon-Avon           1
47                    Conwy                         1
48                    Cornwall                      1
62                    Staffordshire Moorlands       1
68                    South Bucks                   1
Length: 3171, dtype: int64

<h1>Insight 12</h1><h3>The most number of casualty have been recorded in Aberdeen City with 1,181. The district with the least number of casualty is South Bucks with only 1</h3>

In [39]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  object        
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  object        
 5   District Area            660679 non-null  object        
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  object        
 10  Road_Type                660679 non-null  object        
 11  Urban_or_Rural_Area      660679 non-null  object        
 12  Weather_Conditio

<h1>QUESTION 11</h1>
<H3>Number casualties based on light conditions?</H3>

In [40]:
accidents.groupby(['Number_of_Casualties','Light_Conditions']).size()

Number_of_Casualties  Light_Conditions           
1                     Darkness - lighting unknown      5125
                      Darkness - lights lit           98406
                      Darkness - lights unlit          1975
                      Darkness - no lighting          24927
                      Daylight                       373608
                                                      ...  
45                    Daylight                            1
47                    Daylight                            1
48                    Darkness - no lighting              1
62                    Daylight                            1
68                    Darkness - lights lit               1
Length: 88, dtype: int64

<h1>Insight 13</h1><h3>There have been a total of 56% of recorded casualties in Daylight</h3>

In [41]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  object        
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  object        
 5   District Area            660679 non-null  object        
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  object        
 10  Road_Type                660679 non-null  object        
 11  Urban_or_Rural_Area      660679 non-null  object        
 12  Weather_Conditio

<h1>QUESTION 12</h1>
<H3>Number of vehicle in an accident based on light conditions?</H3>

In [42]:
accidents.groupby('Light_Conditions')['Number_of_Vehicles'].mean()

Light_Conditions
Darkness - lighting unknown    1.758482
Darkness - lights lit          1.775498
Darkness - lights unlit        1.752654
Darkness - no lighting         1.599781
Daylight                       1.865385
Name: Number_of_Vehicles, dtype: float64

<h1>INSIGHT 14</h1>
<H3>Number of vehicle is atleast 2 meaning to say most accidents occuring is between to vehicles</H3>

<h1>QUESTION 13</h1>
<H3> How does the number of vehicles involved in an accident relate to the accident severity and the type of vehicle?</H3>

In [43]:
a= accidents.groupby(['Accident_Severity', 'Vehicle_Type'])['Number_of_Vehicles'].mean().unstack()
a.max()

Vehicle_Type
Agricultural vehicle                     1.831509
Bus or coach (17 or more pass seats)     1.857439
Car                                      1.856622
Data missing or out of range             1.666667
Goods 7.5 tonnes mgw and over            1.857820
Goods over 3.5t. and under 7.5t          1.855955
Minibus (8 - 16 passenger seats)         1.825853
Motorcycle 125cc and under               1.840677
Motorcycle 50cc and under                1.905263
Motorcycle over 125cc and up to 500cc    1.876190
Motorcycle over 500cc                    1.860619
Other vehicle                            1.848542
Pedal cycle                              1.848684
Ridden horse                             1.500000
Taxi/Private hire car                    1.864620
Van / Goods 3.5 tonnes mgw or under      1.846426
dtype: float64

<h1>INSIGHT 15</h1>
<H3>A higher number of vehicles on the road(1.8) increases the severity of accidents due to traffic congestion and various influencing factors such as vehicle types and environmental conditions.</H3>

In [44]:
accidents['month'] = accidents['month'].astype('Int64')


In [45]:
accidents['Accident_Severity'] = accidents['Accident_Severity'].fillna('Unknown')


In [46]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                      0
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
year                           0
month                          0
dtype: int64

<h1>QUESTION 14</h1>
<H3>Number of casualties per month and how severe?</H3>

In [47]:

# Group by month and accident severity, summing the number of casualties
casualties_per_month = accidents.groupby(['month', 'Accident_Severity'])['Number_of_Casualties'].sum().reset_index()

# Pivot the table for better readability
table = casualties_per_month.pivot(index='month', columns='Accident_Severity', values='Number_of_Casualties')

# Fill NaN values with 0 for better visualization
table = table.fillna(0)

# Display the table
print(table)


Accident_Severity  Fatal  Serious  Slight
month                                    
1                   1453     9832   59753
2                   1181     9104   56228
3                   1260    10381   61435
4                   1304    10869   58447
5                   1526    11334   63936
6                   1242    11341   63713
7                   1385    11608   65497
8                   1586    11403   62291
9                   1355    11538   63160
10                  1489    11465   67348
11                  1414    11061   68978
12                  1288     9503   59860


In [48]:
grouped.unstack()

NameError: name 'grouped' is not defined

<h1>INSIGHT 16</h1>
<H3>Casualties are highest in May–October, with slight injuries being the most common, while serious injuries peak mid-year and fatalities are highest in August and May, highlighting the need for targeted road safety measures during these months.</H3>

⃣<h1>QUESTION 15
<h3>Which month has the highest total number of casualties, and how does severity vary?

In [ ]:
accidents.groupby(['month', 'Accident_Severity'])['Number_of_Casualties'].sum()


<h1>INSIGHT 17</h1>
<H3>Casualties are highest in May–October, with slight injuries being the most common, while serious injuries peak mid-year and fatalities are highest in August and May, highlighting the need for targeted road safety measures during these months.</H3></H3>

<h1>QUESTION 16</h1>
<H3>How does the number of casualties correlate with weather conditions?</H3>


In [ ]:
accidents.groupby('Weather_Conditions')['Number_of_Casualties'].sum()


<h1>INSIGHTS 18</h1>
<H3>Most accidents occur in fine weather without high winds (703,445), followed by rainy conditions without high winds (112,229), suggesting that driver behavior, rather than severe weather, plays a major role in accident occurrence.</H3>


<h1>QUESTION 17</h1>
<H3>Are urban or rural areas more prone to severe accidents?</H3>


In [ ]:
accidents.groupby(['Urban_or_Rural_Area', 'Accident_Severity'])['Number_of_Casualties'].sum()


<h1>INSIGHT 19</h1>
<H3>Accidents in urban areas result in the highest number of casualties overall, while fatal accidents are more common in rural areas, suggesting that higher traffic density increases minor accidents in cities, whereas rural roads contribute to more severe crashes.</H3>


<h1>QUESTION 18</h1>
<H3>What is the average number of vehicles involved in accidents by severity?</H3>


In [ ]:
accidents.groupby('Accident_Severity')['Number_of_Vehicles'].mean()


<h1>INSIGHT 20</h1>
<H3>On average, slight accidents involve the most vehicles (1.86 per accident), followed by fatal (1.79) and serious (1.68) accidents, indicating that multi-vehicle collisions are more common in minor accidents, while severe crashes may involve single vehicles at higher impact.?</H3>


<h1>QUESTION 19</h1>
<H3> Which road types are associated with the highest number of fatalities?</H3>


In [ ]:
accidents.groupby(['Road_Type', 'Accident_Severity'])['Number_of_Casualties'].sum().loc[:, 'Fatal']


<h1>INSIGHT 21</h1>
<H3>Most accidents occur on single carriageways (12,423), followed by dual carriageways (3,598), indicating that roads with two-way traffic and mixed vehicle speeds pose a higher risk for severe accidents.</H3>


<h1>QUESTION 20</h1>
<H3>Find the Correlation Between Number of Casualties and Other Numeric Factors</H3>


In [ ]:
accidents[['Number_of_Casualties', 'Number_of_Vehicles', 'year', 'month']].corr()


<h1>INSIGHT 21</h1>
<H3>The number of casualties has a weak positive correlation (0.23) with the number of vehicles involved, suggesting that accidents with more vehicles tend to result in more casualties, but the impact is not strong. Additionally, year and month have almost no correlation with casualties or vehicles, indicating that **accident severity remains relatively consistent over time.</H3>
